### PR TITLE
store/tikv: handle new region metas inside StaleEpoch error

### PR DESCRIPTION
**Pull request by disksing**
For https://github.com/pingcap/tikv/issues/974
@ngaut @siddontang

### CODE REVIEW

**Comment by shenli** [comment]
@disksing CI failed

**Comment by disksing** [comment]
Wait for me to find an approach to test :)

**Comment by disksing** [comment]
PTAL @ngaut @shenli

File: store/tikv/mock-tikv/rpc.go [hunk 1]
```diff
@@ -283,6 +291,20 @@ func convertToPbPairs(pairs []Pair) []*kvrpcpb.KvPair {
 	return kvPairs
 }
 
+func encodeRegionKey(r *metapb.Region) *metapb.Region {
+	if r.StartKey != nil {
+		var b []byte
+		b = codec.EncodeBytes(b, r.StartKey)
```
Comment (ngaut) [inline store/tikv/mock-tikv/rpc.go]: b := codec.EncodeBytes(b, r.StartKey)

File: store/tikv/mock-tikv/rpc.go [hunk 2]
```diff
@@ -283,6 +291,20 @@ func convertToPbPairs(pairs []Pair) []*kvrpcpb.KvPair {
...
+		r.StartKey = b
+	}
+	if r.EndKey != nil {
+		var b []byte
+		b = codec.EncodeBytes(b, r.EndKey)
```
Comment (ngaut) [inline store/tikv/mock-tikv/rpc.go]: b := codec.EncodeBytes(b, r.EndKey)

File: store/tikv/region_cache.go [hunk 3]
```diff
@@ -246,6 +246,36 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
...
+	c.dropRegionFromCache(old.VerID())
+
+	for _, meta := range newRegions {
+		if err := decodeRegionMetaKey(meta); err != nil {
+			log.Errorf("newRegion's range key is not encoded: %v, %v", meta, err)
+			continue
+		}
+		for i := range meta.Peers {
+			if meta.Peers[i].GetStoreId() == old.peer.GetStoreId() {
+				meta.Peers[0], meta.Peers[i] = meta.Peers[i], meta.Peers[0]
```
Comment (ngaut) [inline store/tikv/region_cache.go]: I didn't get it, why swap the two peers ?

File: store/tikv/mock-tikv/rpc.go (hunk 1 above)
Comment (disksing) [inline store/tikv/mock-tikv/rpc.go]: `b` is also used as parameter.

File: store/tikv/region_cache.go (hunk 3 above)
Comment (disksing) [inline store/tikv/region_cache.go]: I'll add a note for it.

File: store/tikv/split_test.go [hunk 4]
```diff
@@ -71,3 +71,30 @@ func (s *testSplitSuite) TestSplitBatchGet(c *C) {
...
+	txn := s.begin(c)
+	err = txn.Set([]byte("a"), []byte("a"))
+	c.Assert(err, IsNil)
+	err = txn.Set([]byte("c"), []byte("c"))
+	c.Assert(err, IsNil)
+	err = txn.Commit()
+	c.Assert(err, IsNil)
+
+	// Initiate a split and disable PD client. Now we can only update new
+	// region from kvrpc.
```
Comment (queenypingcap) [inline store/tikv/split_test.go]: Initiate a split and disable the PD client. If it still works, the new region is updated from kvrpc.

**Comment by queenypingcap** [comment]
LGTM

**Comment by disksing** [comment]
PTAL @ngaut

File: store/tikv/region_cache.go [hunk 5]
```diff
@@ -246,7 +241,43 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
...
+	defer c.mu.Unlock()
+
+	c.dropRegionFromCache(old.VerID())
+
+	for _, meta := range newRegions {
+		if err := decodeRegionMetaKey(meta); err != nil {
+			log.Errorf("newRegion's range key is not encoded: %v, %v", meta, err)
+			continue
+		}
+		reorderRegionPeers(meta, old.peer.GetStoreId())
```
Comment (ngaut) [inline store/tikv/region_cache.go]: s/reorderRegionPeers/refreshRegionLeader/

File: store/tikv/region_cache.go [hunk 6]
```diff
@@ -246,7 +241,43 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
...
+	}
+}
+
+// reorderRegionPeers moves the leader peer to the first and makes it easier to
+// try the next peer if the current peer does not respond.
+func reorderRegionPeers(r *metapb.Region, leaderStoreID uint64) {
+	for i := range r.Peers {
+		if r.Peers[i].GetStoreId() == leaderStoreID {
+			r.Peers[0], r.Peers[i] = r.Peers[i], r.Peers[0]
+			break
```
Comment (ngaut) [inline store/tikv/region_cache.go]: s/break/return/

**Comment by ngaut** [comment]
PTAL @coocood

File: store/tikv/region_cache.go [hunk 7]
```diff
@@ -246,7 +241,43 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
...
+// OnRegionStale removes the old region and inserts new regions into the cache.
+func (c *RegionCache) OnRegionStale(old *Region, newRegions []*metapb.Region) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.dropRegionFromCache(old.VerID())
+
+	for _, meta := range newRegions {
+		if err := decodeRegionMetaKey(meta); err != nil {
+			log.Errorf("newRegion's range key is not encoded: %v, %v", meta, err)
```
Comment (coocood) [inline store/tikv/region_cache.go]: I think we should return this error.

File: store/tikv/region_cache.go [hunk 8]
```diff
@@ -246,7 +241,43 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
...
+			continue
+		}
+		reorderRegionPeers(meta, old.peer.GetStoreId())
+		peer := meta.Peers[0]
+		if peer.GetStoreId() != old.peer.GetStoreId() {
```
Comment (coocood) [inline store/tikv/region_cache.go]: This logic is not clear, please add comments about it.

File: store/tikv/region_cache.go [hunk 9]
```diff
@@ -223,12 +223,7 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
 		}
 		// Move leader to the first.
 		if leader != nil {
-			for i := range meta.Peers {
-				if meta.Peers[i].GetId() == leader.GetId() {
-					meta.Peers[0], meta.Peers[i] = meta.Peers[i], meta.Peers[0]
-					break
-				}
-			}
+			reorderRegionPeers(meta, leader.GetStoreId())
```
Comment (coocood) [inline store/tikv/region_cache.go]: Why not name it `moveLeaderToFirst` directly?

File: store/tikv/region_cache.go [hunk 10]
```diff
@@ -246,7 +241,43 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
...
+			meta: meta,
+			peer: peer,
+			addr: old.GetAddress(),
+		})
+	}
+}
+
+// reorderRegionPeers moves the leader peer to the first and makes it easier to
+// try the next peer if the current peer does not respond.
+func reorderRegionPeers(r *metapb.Region, leaderStoreID uint64) {
```
Comment (coocood) [inline store/tikv/region_cache.go]: Will the leaderStoreID always found in peers?
If not, should we returns a bool indicates if leaderStoreID has been moved to first?

Comment (disksing) [inline store/tikv/region_cache.go]: We can always try one by one start from the 1st. If it is not the leader, it will send the real leader back.

**Comment by disksing** [comment]
PTAL @ngaut @coocood

File: store/tikv/region_cache.go [hunk 11]
```diff
@@ -246,7 +241,45 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
...
+	defer c.mu.Unlock()
+
+	c.dropRegionFromCache(old.VerID())
+
+	for _, meta := range newRegions {
+		if err := decodeRegionMetaKey(meta); err != nil {
+			return errors.Errorf("newRegion's range key is not encoded: %v, %v", meta, err)
+		}
+		moveLeaderToFirst(meta, old.peer.GetStoreId())
+		peer := meta.Peers[0]
```
Comment (ngaut) [inline store/tikv/region_cache.go]: s/peer/leader/

File: store/tikv/store_test.go [hunk 12]
```diff
@@ -252,3 +254,53 @@ func (c *busyClient) SendCopReq(addr string, req *coprocessor.Request, timeout t
...
+	defer c.Unlock()
+	c.stop = true
+}
+
+func (c *mockPDClient) GetTS() (int64, int64, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	if c.stop {
+		return 0, 0, errors.New("stopped")
```
Comment (ngaut) [inline store/tikv/store_test.go]: Please define a new error ErrorStopped instead of creating a New error each time.

**Comment by ngaut** [comment]
Rest LGTM.

**Comment by ngaut** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**